acceptancetests/assess_persistent_storage.py

```diff
@@ -1,139 +1,428 @@
-                'Expected token "{}" not found in stored results.'.format(
-                    expected_name))
-        if stored_value != expected_value:
-            raise JujuAssertionError(
-                'Token values do not match. Expected: {} got: {}'.format(
-                    expected_value, stored_value))
 
-
-def get_stored_token_content(client):
-    """Retrieve token values from file stored on the unit.
-
-    Token values are retrieved from the storage units and logged into a file on
-    the application unit.
-
-    :param client: ModelClient object to query.
-    :return: Dict containing 'token name' -> 'token value'.
-    """
-    try:
-        # TODO: Need to pass in application/unit
-        contents = client.get_juju_output(
-            'ssh', ('dummy-storage/0', 'cat', '/tmp/status'))
-    except CalledProcessError as e:
-        raise JujuAssertionError('Failed to read token file: {}'.format(e))
-
-    return {token_name: token_value for token_name, token_value in (
-        line.split(':', 1) for line in contents.splitlines() if line != '')}
+    single_blk_id = \
+        assess_charm_removal_single_block_and_filesystem_storage(client)
+    charm_name = 'dummy-storage'
+    charm_path = local_charm_path(
+        charm=charm_name, juju_ver=client.version)
+    # Run juju deploy dummy-storage --attach-storage <single_blk_id>
+    # Note: --attach-storage has the issue of Bug 1704105 too:
+    # https://bugs.launchpad.net/juju/+bug/1704105
+    client.get_juju_output(
+        'deploy', charm_path,
+        '--attach-storage', single_blk_id,
+        include_e=False, merge_stderr=True)
+    client.wait_for_started()
+    client.wait_for_workloads()
+    assert_app_status(client, charm_name=charm_name, expected='active')
+    storage_list,\
+    storage_type,\
+    persistent_setting,\
+    pool_storage,\
+    pool_setting,\
+    storage_status = get_storage_property(client, storage_id=single_blk_id)
+    assert_storage_number(storage_list=storage_list, expected=1)
+    assert_single_blk_existence(
+        storage_list=storage_list, storage_id=single_blk_id)
+    assert_persistent_setting(
+        storage_id=single_blk_id, found=persistent_setting, expected=True)
+    assert_storage_status(
+        found_id=pool_storage, expected_id=single_blk_id,
+        found_status=storage_status, expected_status='attached')
+    # Run juju remove-application dummy-storage
+    client.get_juju_output('remove-application', charm_name, include_e=False)
+    time.sleep(90)
+    # persistent storage single_blk_id should remain after remove-application
+    assert_single_blk_existence(
+        storage_list=storage_list, storage_id=single_blk_id)
+    # Run juju remove-storage <single_blk_id>
+    # Note: remove-storage has the issue of Bug 1704105 too:
+    # https://bugs.launchpad.net/juju/+bug/1704105
+    client.get_juju_output(
+        'remove-storage', single_blk_id, include_e=False, merge_stderr=True)
+    time.sleep(90)
```

Comment:
There is a story behind this. wait_for functions basically execute get status command with until_timeout(). I never tried wait_for() but had experience in until_timeout() before - it failed to get storage list from juju storage after several iterations, maybe the sleep interval between each run was too short. That's the reason I switched to sleep() as until_timeout() is just an accumulation of shorter sleep(). If sleep() fails on substrate due to performance load or other reason, until_timeout() would highly likely fail as well - as we have to set the total waiting time anyway. Also, the exact moment when the storage is actually removed is unknown to Juju (asynchronous) except to call a watcher - according to Andrew, he is going to add it to this storage feature. I had no idea what that is and how to use it. To avoid all these, I'm using sleep() with a reasonable waiting time for now.